docs/architecture/adr-044-lite-client-with-weak-subjectivity.md

```diff
@@ -121,6 +122,136 @@ network usage.
 Check out the formal specification
 [here](https://github.com/tendermint/tendermint/blob/master/docs/spec/consensus/light-client.md).
 
+### Implementation
+
+There are two primary modes of usage right now:
+
+1) Trusted RPC proxy (wrapping multiple RPC clients + verification)
+2) Part of the IBC light client (only verification bit, no RPC) [spec](https://github.com/cosmos/ics/tree/master/spec/ics-002-client-semantics)
+
+First, we'll need something, which will provide us secure headers & validator sets.
+
+```go
+type Provider interface {
+	// 0 - latest
+	GetFullCommit(height int64) (FullCommit, error)
+}
+```
+
+In case of the proxy it will be a `http` provider (wrapping RPC client). For
+IBC, it will be a `ibc` provider, receiving information from IBC transactions.
+
+Once we have the information, we need to verify it.
+
+```go
+type Verifier struct {
+	chainID            string
+	options            TrustOptions
+	lastVerifiedHeight int64
+	logger             log.Logger
+
+	// Already validated, stored locally
+	trusted PersistentProvider
+
+	// New info, like a node rpc, or other import method.
+	sources Provider
+}
+```
+
+Since providers themselves don't know when they have received a new header
+(or may choose to do so upon a request), we must add a new function to
+`Verifier` - `Verify(height int64) error`.
+
+It should also provide `AutoVerify(period)` option to try & verify new headers
+in the background (optional).
+
+**Sequential vs bisecting verifier**
+
+Verifier should use bisection by default, but provide options to choose a
+different mode OR tweak bisection.
+
+```go
+func LinearVerification() Option {
+	return func(v *Verifier) {
+		v.mode = LINEAR
+	}
+}
+
+func BisectingVerification(trustLevel, trustLevelAdj float) Option {
+	if trustLevel > 1 || trustLevel < 1/3 || trustLevelAdj > 1 || trustLevelAdj < 1/3 {
```

Comment:
I don't see where trustLevel is defined

Reply:
they will be defined as constants

```go
const (
	defaultBisectionTrustLevel = 1/3
	defaultBisectionTrustLevelAdj = 1/3
)
```

Reply:
But is this defined as the maximum amount the voting power can change between the HEAD header and the trusted header?

Reply:
`defaultBisectionTrustLevel` - yes

`defaultBisectionTrustLevelAdj` is used when HEAD and the trusted (or pivot) headers are consecutive (i.e. HEAD height - 3, pivot height - 2)

Reply:
https://github.com/tendermint/tendermint/pull/3911/commits/91f66592f3de67ec47b6621a30b2ee9c6a6be27b